src/Functions/h3GetFaces.cpp

```diff
@@ -0,0 +1,100 @@
+#if !defined(ARCADIA_BUILD)
+#    include "config_functions.h"
+#endif
+
+#if USE_H3
+
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/IFunction.h>
+#include <Common/typeid_cast.h>
+#include <common/range.h>
+
+#include <h3api.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+namespace
+{
+
+class FunctionH3GetFaces : public IFunction
+{
+public:
+    static constexpr auto name = "h3GetFaces";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionH3GetFaces>(); }
+
+    std::string getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const auto * arg = arguments[0].get();
+        if (!WhichDataType(arg).isUInt64())
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of argument {} of function {}. Must be UInt64",
+                arg->getName(), 1, getName());
+
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeUInt8>());
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        const auto * col_hindex = arguments[0].column.get();
+
+        auto dst = ColumnArray::create(ColumnUInt8::create());
+        auto & dst_data = dst->getData();
+        auto & dst_offsets = dst->getOffsets();
+        dst_offsets.resize(input_rows_count);
+        auto current_offset = 0;
+
+        for (const auto row : collections::range(0, input_rows_count))
+        {
+            const UInt64 hindex = col_hindex->getUInt(row);
+
+            int max_faces = maxFaceCount(hindex);
+            std::unique_ptr<int> faces(new int(max_faces));
```

Comment:
Lets try to use std::vector<int> faces out of loop, resize it to maxFaceCount inside loop and pass into getIcosahedronFaces faces.data(), it should work much faster because we can avoid allocation inside loop.